### ISSUE TITLE
waf fails to extract ifort compiler version on Windows (repeat of #1473)

### ISSUE THREAD

**Issue by apthorpe** [waf-project/waf#1672]
I'm running into the same problem detailed in (closed) bug #1473 where waf find the Intel Fortran compiler (ifort) on Windows 7, executes it to get the version info, but fails to parse the results, thus claiming it can't find the compiler.

A minimal test case is attached including the build/config.log which shows the proper detection of the compiler binary and the output from running it to get version info.

[winifort_failure_report.zip](https://github.com/waf-project/waf/files/64525/winifort_failure_report.zip)

The important part of log file is at the end:

    ['c:\\Program Files (x86)\\Intel\\Composer XE 2013 SP1\\bin\\ia32\\ifort.exe']
    err: Intel(R) Visual Fortran Compiler XE for applications running on IA-32, Version 14.0.2.176 Build 20140130^M
    Copyright (C) 1985-2014 Intel Corporation.  All rights reserved.^M
    ^M
    ifort: command line error: no files specified; for help type "ifort /help"^M
     
    from C:\Users\Apthorpe\Downloads\inst\winifort-0: could not determine the compiler version ['c:\\Program Files (x86)\\Intel\\Composer XE 2013 SP1\\bin\\ia32\\ifort.exe']
    not found
    from C:\Users\Apthorpe\Downloads\inst\winifort-0: could not configure a Fortran compiler!

I haven't been able to test a fix, but if it turns out that the ifort version detection regex is failing, you may want to change waflib/Tools/ifort.py from

    @conf
    def get_ifort_version(conf,fc):
            version_re=re.compile(r"Intel[\sa-zA-Z()0-9,-]*Version\s*(?P<major>\d*)\.(?P<minor>\d*)",re.I).search

to

    @conf
    def get_ifort_version(conf,fc):
            version_re=re.compile(r"Intel[^A-Z0-9].*Version\s*(?P<major>\d*)\.(?P<minor>\d*)",re.I).search

If the re.I option is specified, there's no need to look for both [a-z] and [A-Z]; the regex can be loosened a little so if Intel randomly sprinkles more punctuation in their difficult-to-gracefully-obtain version info next year, hopefully the regex won't start failing again.

**Comment by ita1024** [comment]
> hopefully the regex won't start failing again

Is there any command that we could use to display the version in the same way as [get_cc_version](https://github.com/waf-project/waf/blob/master/waflib/Tools/c_config.py#L1011)?

**Comment by apthorpe** [comment]
I tend to write fairly precise regexes, which is good for avoiding false positives but it's a liability in this case. Intel seems determined to be inconsistent in ifort's commands and presentation of version info (e.g. ifort on Linux supports --version; ifort on Windows doesn't have an explicit version command). I'm rusty on Python's re syntax and how to specify word boundaries; what might be better than my suggestion above is /\bintel\b.*\bversion\b{remainder to capture major & minor version numbers}/i Matching "blah blah intel blah blah version 1.2.3.4" may be good enough; it's good practice to avoid .* but it may be warranted here.

Anyway, I've spent maybe 4 hours total working with waf; I'll see if I can isolate the problem further & maybe come up with an actual patch.

**Comment by ita1024** [comment]
The question is, does ifort nowadays provide a command such as `"ifort -dM <emptyfile>"` to display macros just like in c++ mode? I can have the detection re-use `get_cc_version` if necessary but i do not have ifort on windows to try it with.

**Comment by apthorpe** [comment]
Further digging and experimentation yielded this:

    C:\>ifort /E /fpp a.f90 /dryrun
    Intel(R) Visual Fortran Compiler XE for applications running on IA-32, Version 14.0.2.176 Build 20140130
    Copyright (C) 1985-2014 Intel Corporation.  All rights reserved.
    ifort: command line warning #10006: ignoring unknown option '/dryrun'
    
    c:\PROGRA~2\Intel\COMPOS~1\bin\ia32\fpp  \
        -D__INTEL_COMPILER=1400 \
        -D__INTEL_COMPILER_UPDATE=2 \
        -D_WIN32 \
        -D_M_IX86=700 \
        -D_MSC_VER=1600 \
        -D_MSC_FULL_VER=160030319 \
        -D_MSC_EXTENSIONS \
        -D_MT \
        -D__INTEL_COMPILER_BUILD_DATE=20140130 \
        -D__INTEL_OFFLOAD \
        -D__SSE2__ \
        -D__SSE__ \
        -D__MMX__ \
        -I. \
        -Ic:\PROGRA~2\Intel\COMPOS~1\compiler\include \
        "-Ic:\Program Files (x86)\Intel\Composer XE 2013 SP1\compiler\include" \
        "-Ic:\Program Files (x86)\Intel\Composer XE 2013 SP1\compiler\include\ia32" \
        "-IC:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\INCLUDE" \
        -I\PlatformSDK\include \
        "-Ic:\Program Files (x86)\Intel\Composer XE 2013 SP1\mkl\include" \
        -Ic:\PROGRA~2\Intel\COMPOS~1\compiler\include\ia32 \
        -free \
        -4Ycpp \
        -4Ncvf \
        -f_com=yes \
        a.f90

This is based on a thread at StackOverflow https://stackoverflow.com/questions/9349754/generate-list-of-preprocessor-macros-defined-by-the-compiler

The equivalent for ifort on linux is `ifort -E -fpp a.f90 -dryrun` except that `-dryrun` doesn't throw a warning.

**Comment by ita1024** [comment]
This is unclear. Does running `ifort -E -fpp a.f90 -dryrun` with an empty `a.f90` file return the defines somewhere at all? If yes, are they in a `a.*` output file or are they in the output, in which case the complete output is missing? In the latter case is that on stdout or stderr?

**Comment by apthorpe** [comment]
To clarify, on linux,

    touch a.f90 ; ifort -E -fpp a.f90 -dryrun 1>aprobe_out.txt 2>aprobe_err.txt; /bin/rm a.f90

leaves `aprobe_out,txt` empty and puts the attached option dump in `aprobe_err.txt`

[aprobe_err.txt](https://github.com/waf-project/waf/files/66295/aprobe_err.txt)

I'll check which stream the Windows version of `ifort`/`fpp` writes to when I'm in the office tomorrow.

**Comment by ita1024** [comment]
So in other words, we cannot get the versions or the macros on Linux at least. Tell me the result on Windows.

**Comment by haraldkl** [comment]
Oh, I don't know why I missed this issue up to now.
In any case. I do not know, whether this is related to the issue on windows or not, but I am seeing trouble on Linux with different compilers. Their common problem is that, they return an error code if they do not get some file to compile (which we attempt mostly when looking for the version string). Namely I see this issue right now with the Intel and the NEC compiler.
This problem was introduced in bfbcc73eb21d19f25ad7ccccc9231522e0a6a897 as far as I can tell, as the cmd_and_log will check for the return code, while previously we were using the subprocesses directly and ignoring error codes.

**Comment by haraldkl** [comment]
I am not aware of a method to obtain defines and the like as we do it in GCC. Thus, I believe it would be nice to have a method to deal with error codes by the compiler without giving up. One more note: It might be that on Linux the change in the returned error code was rather recent. On a Cray system with Intel 15.0.4.233 I get the failure, while on another system with 15.0.1.133 the compiler does not return an error code.

**Comment by haraldkl** [comment]
For intel there are some options to avoid compilation (for example -help or -print-multi-lib). We could use those to avoid the non-zero return codes. But I am not sure, whether we can find such options for all compilers.

**Comment by ita1024** [comment]
Ok, so the regexp change above and disabling the exit code verification will be applied. You can add the changes else i will have a look tonight.

**Comment by haraldkl** [comment]
Thanks a lot for your fast reviews and suggestions, Thomas. My issues are gone with the changes, as far as I can tell. I hope this also resolves problems on Windows systems.

I wish you some nice holidays and a good start into the new year!

**Comment by ita1024** [comment]
@haraldkl Thanks, have nice holidays too!
@apthorpe the issues should be gone, please verify :-)

**Comment by apthorpe** [comment]
@haraldkl @ita1024 Thank you both for your assistance and guidance on testing; I'll check our system and get back to you after the holidays

**Comment by apthorpe** [comment]
I'm not sure if this helps, but I poked at the Intel developer forums and they suggested

    ifort /list a.f90

which, for an empty a.f90, produces the attached human-formatted report of compiler options. I understand that compiler options aren't preprocessor macros or defines but I thought it might help.

[a.lst.txt](https://github.com/waf-project/waf/files/73552/a.lst.txt)

I believe my original specific problem of ifort not being detected has been solved with the most recent fixes. My bigger picture problem remains so it's likely I'll be posting more questions or issues with ifort on Windows. And again, feel free to ask me to run tests and send results. ifort is prohibitively expensive which keeps it from being supported more broadly by open dev tools so unless I contribute help back, nothing's going to improve... Thanks again.
